### PR TITLE
[all] remove all $ globals

### PR DESCRIPTION
**Pull request by hjanuschka**
lets try to get rid of $ variables.
feedback welcome. - i ll split the PR once its reviewed.

### CODE REVIEW

**Comment by milch** [comment]
Is there a good reason for why we need this?

**Comment by KrauseFx** [comment]
There has been a discussion around this somewhere by @hjanuschka, I can't find the issue. In general this is a good idea, since it allows everyone to see what variables are available, and it's easier and cleaner to write tests 👍

**Comment by hjanuschka** [comment]
well i actually think having global variables is not the best thing.
internally we have a penalty box where you put in some € for introducing new globals to some code 👍 

and i brought atleast two into fastlane, `$capture_output`  - `$captured_output`
i think the goal of this PR should be to encapsulate the fastlane global state.


imagine another gem that sets $verbose?!

**Comment by milch** [comment]
Hmm true, hadn't thought about other gems changing the globals. `verbose` is a pretty common name so the chances of that happening aren't even that slim. 

Anyway, there are loads of errors on the tests, if we want to merge this then we'd need to get rid of those first.

**Comment by hjanuschka** [comment]
@milch  going to work on this pr in the next few days. i ll cc you once i get a green test state ;)

**Comment by hjanuschka** [comment]
@milch  tests pass now, i also added a explicit test for the globals class.

BUT i had to revert the changes in `credentials_manager` as it does not depend on fastlane_core. 😢 

let me know if we are ready for a PR split 👍

**Comment by milch** [comment]
> BUT i had to revert the changes in credentials_manager as it does not depend on fastlane_core. 😢

Hmm, that is interesting... Could that present any problems since the globals could be out of sync?

**Comment by hjanuschka** [comment]
@milch i think it would be good to also set the $verbose - what is the reason why we cannot introduce fastlane_core as a dep. to `credentials_manager`  ?

**Comment by milch** [comment]
@hjanuschka 

> i think it would be good to also set the $verbose


> what is the reason why we cannot introduce fastlane_core as a dep. to credentials_manager ?

The reason is that fastlane_core depends on credentials_manager, so we would be introducing a circular dependency. @KrauseFx and I are currently working on #7436 which will solve that, so let's maybe just defer this PR until that is merged.

**Comment by hjanuschka** [comment]
@milch ok lets defer it monogem FTW!

**Comment by hjanuschka** [comment]
fixed merge conflicts, ready for another review round! since monogem, cred_manager should work again.

File: fastlane_core/lib/fastlane_core/globals.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    class << self
+      attr_writer :captured_output
+      attr_writer :capture_output
+      attr_writer :verbose
+    end
+
+    def self.capture_output?
+      return nil unless @capture_output
+      return true
```
Comment (mfurtak) [inline fastlane_core/lib/fastlane_core/globals.rb]: Why have the extra check here? Is there a reason we want to return `nil` when the value is `false`?

File: fastlane_core/lib/fastlane_core/globals.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+module FastlaneCore
+  class Globals
+    def self.captured_output
+      unless @captured_output
+        @captured_output = ""
+      end
+      @captured_output
```
Comment (mfurtak) [inline fastlane_core/lib/fastlane_core/globals.rb]: Is there a reason we need to assign `@captured_output = ""` as part of this? Would 

```ruby
def self.captured_output
  @captured_output || ""
end
```

be OK?

File: fastlane_core/lib/fastlane_core/globals.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    def self.capture_output?
+      return nil unless @capture_output
+      return true
+    end
+
+    def self.captured_output?
+      if @capture_output && @captured_output.length > 0
+        return true
+      end
+      return false
```
Comment (mfurtak) [inline fastlane_core/lib/fastlane_core/globals.rb]: Is there a need to conditional checks and return boolean values directly?

```ruby
def self.captured_output?
  @capture_output && @captured_output.length > 0
end
```

File: fastlane_core/lib/fastlane_core/globals.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    end
+
+    def self.verbose?
+      return nil unless @verbose
+      return true
```
Comment (mfurtak) [inline fastlane_core/lib/fastlane_core/globals.rb]: Why have the extra check here? Is there a reason we want to return nil when the value is false?

**Comment by mfurtak** [comment]
I believe that _spaceship_ shouldn't be made to depend on _fastlane_core_, so that part of this refactor is problematic. /cc @KrauseFx

**Comment by hjanuschka** [comment]
@mfurtak @KrauseFx  how about checking if fastlane-core is loaded/available and then use the global class, and if not found/available - fallback to `$verbose` ?

if we drop the globals class in spaceship (and keep using `$` ) - i think we have a similar situation as on the "before-monogem" time - where we had troubles with `credentials_manager`  and deferred the PR.

btw.: once we have a decision on how to deal with spaceship - i am going to work on the open comments 👍

**Comment by DanToml** [comment]
@hjanuschka that seems like it should work to me :+1:

File: fastlane_core/lib/fastlane_core/globals.rb (hunk 2 above)
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/globals.rb]: or 
`@captured_output ||= ""`
if you need to assign it

Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/globals.rb]: thx fixed

File: fastlane_core/lib/fastlane_core/globals.rb (hunk 1 above)
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/globals.rb]: there are many places where it is explcitily required to be nil, and `false`  would change the behaviour.

File: fastlane_core/lib/fastlane_core/globals.rb (hunk 3 above)
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/globals.rb]: thx!! thats awesome.

File: fastlane_core/lib/fastlane_core/globals.rb (hunk 4 above)
Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/globals.rb]: same as above

**Review by hjanuschka — COMMENTED** [review]
@DanToml @mfurtak feedback addressed. spaceship is no indpendant of fastlane-core, but if it is loaded, it will respect the verbose state from fastlane-core.

master rebased, conflicts solved.
really looking forward to get this done.

File: fastlane_core/lib/fastlane_core/globals.rb (hunk 4 above)
Comment (DanToml) [inline fastlane_core/lib/fastlane_core/globals.rb]: This seems bad, but outside the scope of this or to fix. This is something that we should fix tho

Comment (hjanuschka) [inline fastlane_core/lib/fastlane_core/globals.rb]: yeah but i ll preferr to do this in a seperate PR - once this is merge, return a real false, and then fix the broken things (mostly test cases that are failing) - i just want to keep the mistake-surface as low as possible - as this PR already changes a lot.

**Review by endocrimes — APPROVED** [review]

**Review by KrauseFx — APPROVED** [review]
Exciting 👍

File: spaceship/lib/spaceship/globals.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+module Spaceship
+  class Globals
+    # if spaceship is run with a FastlaneCore available respect the global state there
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/globals.rb]: Nice 👍

**Pull request merged**